[WPE] Implement click counting in mouse events

dpino opened the pull request:
#### 0482f3984fab2efcfa7a4d32574c61ad144e27d0
<pre>
[WPE] Implement click counting in mouse events
<a href="https://bugs.webkit.org/show_bug.cgi?id=174674">https://bugs.webkit.org/show_bug.cgi?id=174674</a>

Reviewed by Carlos Garcia Campos.

Follow WinCairo logic to implement click counts for mouse events.

* LayoutTests/platform/wpe/TestExpectations:
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::getDoubleClickTime):
(WebKit::clickCount):
(WebKit::WebEventFactory::createWebMouseEvent):

Canonical link: <a href="https://commits.webkit.org/265808@main">https://commits.webkit.org/265808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7062e8c46b7c6d177c1086f6696e56557cf4008b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11960 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12611 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13709 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11544 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12324 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14246 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12124 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10121 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14121 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10235 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10857 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/11313 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11016 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14184 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11497 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10711 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2918 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15038 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11354 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -157,6 +157,41 @@ static inline short pressedMouseButtons(uint32_t modifiers)
+// To fix it, it would be needed that libwpe provides milliseconds information in
+// wpe_input_pointer_event. Right now it only provides time in seconds.
+// See: https://github.com/WebPlatformForEmbedded/libwpe/blob/master/include/wpe/input.h#L79
+static inline double getDoubleClickTime()
```
The time is an uint32_t but it should be milliseconds.

dpino replied:
I think `time` in `wpe_input_pointer_evet` stores UNIX timestamps, which is seconds.

From `WebEventFactory.cpp`:

```
// WPE event time field is an uint32_t, too small for full ms timestamps since
// the epoch, and are expected to be just timestamps with monotonic behavior
// to be compared among themselves
```

https://searchfox.org/wubkat/source/Source/WebKit/Shared/libwpe/WebEventFactory.cpp#95

If I make `getDoubleClickTime()` return 500, which should work if `time` is milliseconds, the two tests that are passing now are failing.

I think it would be useful to store milliseconds, microseconds or nanoseconds information as part of `wpe_input_pointer_evet`, probably as an extra field.

carlosgcampos replied:
That's only true for tests, it's a bug in EventSenderProxyLibWPE, I would say, that uses seconds for the synthesized events, but it's not the case for actual events.

dpino replied:
Got it. Actually the tests still pass if I made `getDoubleClickTime()` return 500.

[EventSenderProxy.h:m_time](https://searchfox.org/wubkat/source/Tools/WebKitTestRunner/EventSenderProxy.h#144) stores time as a double. From what I could see while tracing the code it seems `m_time` stores uptime of the machine (seconds plus decimal part for nanoseconds). `EventSenderProxyLibWPE` is working fine, it converts `m_time` to a milliseconds timestamp.

https://searchfox.org/wubkat/source/Tools/WebKitTestRunner/libwpe/EventSenderProxyLibWPE.cpp#146

carlosgcampos commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -157,6 +157,37 @@ static inline short pressedMouseButtons(uint32_t modifiers)
+static inline uint32_t getDoubleClickTime()
+{
+    return 500; // Milliseconds.
+}
```
Why do we need a function for this? it can be a static const, and the pixelThreshold too.

carlosgcampos commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -157,6 +157,37 @@ static inline short pressedMouseButtons(uint32_t modifiers)
+
+static inline unsigned clickCount(struct wpe_input_pointer_event* event)
+{
+    const unsigned pixelThreshold = 5;
```
This could be static const

dpino committed:
[WPE] Implement click counting in mouse events
https://bugs.webkit.org/show_bug.cgi?id=174674

Reviewed by Carlos Garcia Campos.

Follow WinCairo logic to implement click counts for mouse events.

* LayoutTests/platform/wpe/TestExpectations:
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::getDoubleClickTime):
(WebKit::clickCount):
(WebKit::WebEventFactory::createWebMouseEvent):

Canonical link: https://commits.webkit.org/265808@main
